In [87]:
references = [
    "https://ts.hust.edu.vn/training-cate/nganh-dao-tao-dai-hoc/dien-tu-va-vien-thong",
    "https://huongnghiep.hocmai.vn/chuyen-nganh/ky-thuat-dien-tu-vien-thong/dai-hoc-bach-khoa-ha-noi/",
    "https://blog.jobsnew.vn/nganh-dien-tu-vien-thong/"
]

In [88]:
hypothesis = "# Báo cáo về Ngành Điện tử - Viễn thông của Đại học Bách khoa Hà Nội \
## Giới thiệu chung \
Trong bối cảnh cuộc Cách mạng công nghiệp lần thứ 4 đang diễn ra mạnh mẽ, ngành Điện tử - Viễn thông trở thành một trong những ngành học quan trọng và tiềm năng, đóng vai trò chủ chốt trong sự phát triển của xã hội hiện đại. Đại học Bách khoa Hà Nội (HUST), một trong những trường đại học hàng đầu Việt Nam, đã và đang đào tạo ngành Điện tử - Viễn thông, nhằm cung cấp nguồn nhân lực chất lượng cao cho thị trường lao động trong và ngoài nước. \
## Chương trình đào tạo \
Chương trình đào tạo ngành Điện tử - Viễn thông tại HUST được thiết kế để trang bị cho sinh viên những kiến thức toàn diện từ cơ bản đến chuyên sâu. Sinh viên sẽ được học các khối kiến thức toán học và khoa học cơ bản, kiến thức cơ sở ngành Điện tử - Viễn thông, kiến thức bổ trợ xã hội và kỹ năng mềm, cũng như khối kiến thức tự chọn chuyên sâu theo các định hướng như kỹ thuật điện tử, kỹ thuật máy tính, kỹ thuật thông tin truyền thông, kỹ thuật y sinh, kỹ thuật điện tử hàng không vũ trụ, kỹ thuật đa phương tiện, kỹ thuật vi điện tử và công nghệ nano (HUST, n.d.). \
Sinh viên có cơ hội lựa chọn một trong các chuyên ngành sau: \
- Điện tử và kỹ thuật máy tính: chuyên về lập trình cho các hệ thống nhúng. \
- Kỹ thuật thông tin - truyền thông: chuyên về hệ thống viễn thông, kỹ thuật truyền số liệu. \
- Kỹ thuật y sinh: chuyên nghiên cứu các thiết bị y tế phục vụ khám chữa bệnh. \
- Kỹ thuật điện tử hàng không vũ trụ: nghiên cứu về các loại máy bay tự hành. \
- Kỹ thuật Đa phương tiện: nghiên cứu về các thiết bị nghe nhìn, điều chỉnh âm độ của các thiết bị thu âm (Hocmai.vn, n.d.). \
## Tiêu chuẩn đầu vào và cơ hội chuyển tiếp \
Điểm chuẩn ngành Kỹ thuật Điện tử - Viễn thông tại HUST thường nằm ở mức cao, phản ánh sự cạnh tranh và chất lượng đào tạo của ngành. Ví dụ, điểm chuẩn năm 2023 qua các phương thức xét tuyển khác nhau như Đánh giá tư duy và kết quả thi tốt nghiệp THPT dao động từ 24.5 đến 26.8 (Hocmai.vn, n.d.). \
Ngoài ra, sinh viên còn có cơ hội chuyển tiếp sang các chương trình đào tạo quốc tế, như chương trình liên kết với Đại học Leibniz Hannover (Đức), với giai đoạn đầu học tại HUST và có cơ hội chuyển tiếp sang Đức để học bằng thạc sĩ (HUST, n.d.). \
## Cơ hội nghề nghiệp \
Sau khi tốt nghiệp, sinh viên ngành Điện tử - Viễn thông của HUST có thể làm việc tại các hãng thiết bị, các tập đoàn, công ty đa quốc gia về Điện tử - Viễn thông. Theo thống kê, 95% sinh viên khoa Kỹ thuật Điện tử - Viễn thông của HUST có việc làm ngay sau một năm tốt nghiệp với mức lương khởi điểm từ 8-20 triệu đồng (Hocmai.vn, n.d.). \
Các vị trí công việc sau khi tốt nghiệp bao gồm nhưng không giới hạn ở: \
- Chuyên viên quản lý kỹ thuật và sản phẩm, quản lý dự án kinh doanh. \
- Chuyên viên tư vấn, thiết kế và vận hành kỹ thuật tại các đài phát thanh, đài truyền hình, công ty tư vấn thiết kế mạng viễn thông, sản xuất thiết bị Điện tử - Viễn Thông. \
- Nghiên cứu và phát triển, mạng vô tuyến, định vị dẫn đường, thiết bị tín hiệu đa phương tiện (Jobsnew.vn, n.d.). \
## Kết luận \
Ngành Điện tử - Viễn thông tại Đại học Bách khoa Hà Nội không chỉ cung cấp một nền tảng kiến thức vững chắc mà còn mở ra cơ hội nghề nghiệp rộng lớn trong một lĩnh vực đang phát triển mạnh mẽ. Với chương trình đào tạo chất lượng cao, tiêu chuẩn đầu vào nghiêm ngặt và cơ hội chuyển tiếp quốc tế, sinh viên tốt nghiệp từ HUST có thể tự tin bước vào thị trường lao động và đóng góp vào sự phát triển của ngành công nghiệp Điện tử - Viễn thông cả trong và ngoài nước."

In [89]:
import urllib3
urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

In [90]:
import requests
from bs4 import BeautifulSoup


class BaseLoader:
    def __init__(self, urls):
        self.urls = urls

    def load(self):
        references = []
        for url in self.urls:
            response = requests.get(url, verify=False)
            soup = BeautifulSoup(response.text, 'html.parser')
            references.append(soup.get_text())
        
        return references
            
loader = BaseLoader(references)
references = loader.load()

In [91]:
# Normalizing Vietnamese text
import unicodedata
import re

def normalize_text(text):
    text = unicodedata.normalize("NFC", text)
    text = re.sub(r"\s+", " ", text)
    return text

In [92]:
references = [normalize_text(content) for content in references]
hypothesis = [normalize_text(hypothesis)]

In [93]:
if not references:
    print('The references list is empty.')
else:
    reference = ' '.join(references)
    with open('refs.txt', 'w') as f:
        f.write(reference)

In [94]:
if not hypothesis:
    print('The hypothesis list is empty.')
else:
    with open('hypos.txt', 'w') as f:
        for hypothesis in hypothesis:
            f.write(hypothesis + "\n")

In [95]:
from bert_score import score

In [96]:
with open("hypos.txt") as f:
    cands = [line.strip() for line in f]

with open("refs.txt") as f:
    refs = [line.strip() for line in f]

In [97]:
# Hide the loading messages
import logging
import transformers
transformers.tokenization_utils.logger.setLevel(logging.ERROR)
transformers.configuration_utils.logger.setLevel(logging.ERROR)
transformers.modeling_utils.logger.setLevel(logging.ERROR)

In [98]:
P, R, F1 = score(cands, refs, lang="vi", verbose=False)

In [99]:
print(f"Precision: {P.mean():.3f}")
print(f"Recall: {R.mean():.3f}")
print(f"F1 score: {F1.mean():.3f}")

Precision: 0.717
Recall: 0.677
F1 score: 0.696


#### F1 score is 0.696 that is quite good for my project!!!

#### Using ROUGE score to evaluate the quality of the generation text.

In [100]:
# Now, let's define the function to calculate Rouge score with the refs.txt and hypos.txt files
from rouge_score import rouge_scorer

def rouge_score(hypothesis_file, reference_file):
    scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)
    with open(reference_file) as f:
        references = [line.strip() for line in f]
    with open(hypothesis_file) as f:
        hypothesis = [line.strip() for line in f]
    scores = [scorer.score(ref, hypo) for ref, hypo in zip(references, hypothesis)]
    return scores

rouge = rouge_score("hypos.txt", "refs.txt")

In [101]:
for i, score in enumerate(rouge):
    print(f"Rouge-1 precision: {score['rouge1'].precision:.3f}")
    print(f"Rouge-2 precision: {score['rouge2'].precision:.3f}")
    print(f"Rouge-L precision: {score['rougeL'].precision:.3f}")

Rouge-1 precision: 0.996
Rouge-2 precision: 0.925
Rouge-L precision: 0.803


#### The ROUGE score is used to evaluate the quality of the text summarization system. The ROUGE score is calculated based on the n-gram overlap between the generated text and the reference text. In this project, the generated text is a summary of the reference text, so the ROUGE score is suitable for this project. For my project, I highly tend to use the precision score of ROUGE because it is the most important score to evaluate the quality of the generation text.

1. Precision: Precision is the number of overlapping n-grams between the generated text and the reference text divided by the number of n-grams in the generated text. Precision is used to evaluate the quality of the generated text. The higher the precision, the better the quality of the generated text.
2. Recall: Recall is the number of overlapping n-grams between the generated text and the reference text divided by the number of n-grams in the reference text. Recall is used to evaluate the coverage of the generated text. The higher the recall, the better the coverage of the generated text.
3. F1: F1 is the harmonic mean of precision and recall. F1 is used to evaluate the balance between precision and recall. The higher the F1, the better the balance between precision and recall.